### PR TITLE
[runtime] Binary section data loading for extra ELF images

### PR DESCRIPTION
**Pull request by spevans**
Use static constructors for registering dlopen-ed images with the runtime.

- For ELF targets, add ImageInspectionInit.cpp that wil be prepended
  to each ELF shared object to automatically add the section data
  as the object is dynamically loaded (rdar://problem/19045112)

### CODE REVIEW

**Comment by spevans** [comment]
@jckarter couple of notes:

I had another go at adding a struct SectionInfo into the swift_section.S but this time I added it as a new symbol and didnt change the existing ones. This meant I could get rid of the getSectionInfo(). If this isnt ideal I have another version keeping the old method

I couldnt work out what would happen if a 2nd thread dlopend an image while dl_iterate_phdr() was in progress so I just removed all of the dlopen()/dl_iterate_phdr() and just used the constructors. This should have the advantage of eliminating lots of callbacks for images we arent interested in and also saves on dlsym() lookups

**Comment by jrose-apple** [comment]
:-( We're trying very hard to *avoid* static constructors. I don't think this is worth the tradeoff if a particular program never asks for conformances.

**Comment by jckarter** [comment]
Static constructors are the only public interface Linux and Windows' dynamic loader have for load-time hooks, so I think it's fine in this case, though we should be careful not to do any more work than absolutely necessary to register the new image with the runtime.

**Comment by jckarter** [comment]
(At the ntdll level, Windows does have `LdrRegisterDllNotification`, but that's not documented or guaranteed to be supported in future versions.)

File: stdlib/public/runtime/ImageInspection.h [hunk 1]
```diff
@@ -30,6 +30,23 @@ namespace swift {
...
+
+// These are defined in swift_sections.S
+extern "C" const SectionInfo __swift2_protocol_conformances_block;
+extern "C" const SectionInfo __swift2_type_metadata_block;
+
+// Called by injected constructors when a dynamic library is loaded
+void addImageProtocolConformanceBlock(const SectionInfo *block);
+void addImageTypeMetadataRecordBlock(const SectionInfo *block);
+
+#endif // defined(__ELF__) || defined(__ANDROID__)
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspection.h]: I don't think this is worth changing. We're wasting space for pointers that can be trivially recovered.

File: stdlib/public/runtime/ImageInspectionInit.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+using namespace swift;
+
+// This is called at startup and by each shared object as it is dlopen()'d to
+// allow the section data for the object to be loaded
+__attribute__((constructor))
+static void
+loadSectionData() {
+  addImageProtocolConformanceBlock(&__swift2_protocol_conformances_block);
+  addImageTypeMetadataRecordBlock(&__swift2_type_metadata_block);
+}
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspectionInit.cpp]: This is doing too much work, since it will force initialization of the lookup caches even if they haven't been enabled yet, and is not robust ABI if `libswiftCore.so` changes and requires new per-image registration hooks in the future. It would be better for the runtime to export an entry point `swift_registerLoadedImage(const void *image)`, and have the static constructor do nothing but call that with a handle to the loaded image, something like:

```
SWIFT_RUNTIME_EXPORT extern "C" void swift_registerLoadedImage(const void *dsoHandle);

__attribute__((constructor))
static void onLoad() {
  swift_registerLoadedImage(__dso_handle);
}
```

The runtime implementation of `swift_registerLoadedImage` should be responsible for extracting and registering the sections that parts of the runtime are interested in, and do so only if those parts have already been initialized. It should be as close to a no-op as possible if an image is loaded before any lookup tables are initialized.

We should also make sure that this static constructor is not linked into executables or into `libswiftCore.so` itself, since the executable and runtime by their nature have to have been loaded before any lookup caches are triggered.

**Review by jckarter — CHANGES_REQUESTED** [review]

**Comment by jckarter** [comment]
Another approach that would maximize laziness and avoid the need for a static constructor, at the cost of making failed lookups a bit more expensive, would be to have the lookup caches fall back to re-iterating all loaded images using `dl_iterate_phdr` if a lookup doesn't get a cache hit after processing all currently-registered sections. I'm not sure how expensive the `dl_iterate_phdr` iteration is. The approach could be paired with a simplified static constructor that just sets a `loadedNewImages` flag in the runtime, which failed conformance lookup would check before re-iterating the images.

**Comment by spevans** [comment]
@jckarter I updated to remove the SectionInfo that was added to the swift_sections.S

I also renamed some of the functions used by the static init construtor, they may have implied that the caches were being initialised at startup but the lazy initialisation is still being used. The static init construtor just saves the SectionInfo to a list and its not used until ```initializeProtocolConformanceLookup()``` or ```initializeTypeMetadataRecordLookup()``` is called as is currently the case.

This also has the advantage that we dont need to avoid linking the static construtor into ```libswiftCore.so```

I had a look at glibc but I couldnt see any export of information about images loaded by ```dlopen```. Also, calling ```dl_iterate_phdr``` is quite expensive as a lock is held for its entire duration to stop any other images being loaded and ```dlsym``` is also quite complex, so I think this current solution does the least work while also being sufficiently lazy.

**Comment by jckarter** [comment]
This would still hardcode an exact set of operations into every dynamic library, which would prevent `libswiftCore.so` from being upgraded to a version that needs additional per-image hooks. I really think that the static constructor should be either a single call to a function provided by the runtime, so that the runtime has full control of what happens. Organization-wise, that implementation should be able to live entirely in `ImageInspectionELF.cpp` so that we don't need to export its details out of the runtime, beyond a single public entry point.

**Comment by spevans** [comment]
Ive simplified the static constructor to just pass an address in the DSO to ```addNewDSOImage``` in ```ImageInspectionELF.cpp```. When the caches are later initialised this address can be used to find the image filename using ```dladdr``` and the section data loaded via ```dlopen/dlsym```. This keeps all logic in ```ImageInspectionELF.cpp```

File: stdlib/public/runtime/ImageInspectionELF.cpp [hunk 3]
```diff
@@ -35,67 +39,116 @@ static const char ProtocolConformancesSymbol[] =
...
+      typeMetadataLock = new Mutex();
+      typeMetadataRecordDSOAddrs = new vector<const void *>();
+
+      // Add the main executable to the lists
+      protocolConformanceDSOAddrs->push_back(nullptr);
+      typeMetadataRecordDSOAddrs->push_back(nullptr);
+    }, nullptr);
+
+  lazyInitProtocolConformance(addr);
+  lazyInitTypeMetadataRecord(addr);
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: Code organization looks much better, thanks! Keep in mind though that `__attribute__((constructor))`s fire not only at `dlopen` time, but at process load time, so all this laziness will be wasted when `addNewDSOImage` gets invoked early at process start. We need to keep track of whether each cache has been initialized to avoid doing any work if it hasn't. Something like this:

```
static std::atomic<bool> didInitializeProtocolConformanceLookup;
void swift::initializeProtocolConformanceLookup() {
  // Set `didInitializeProtocolConformanceLookup` so that loading
  // future Swift images will register with the lookup cache.
  didInitializeProtocolConformanceLookup.store(true, std::memory_order_release);
  // Search the loaded dls...
  ...
}

extern "C"
void swift::swift_addNewDSOImage(const void *addr) {
  if (didInitializeProtocolConformanceLookup.load(std::memory_order_acquire)) {
    addImageProtocolConformanceBlockCallback(...get section from addr...);
  }
}
```

and a similar pattern for the type metadata lookup cache.

Comment (spevans) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: The code actually relies on ```addNewDSOImage``` being called before either cache is initialised as it creates a Mutex and array for each cache. Then ```addNewDSOImage``` just stores an address in the DSO into each list. When a cache is initialised, the addresses in that cache's list are used to load the section data from the DSO, then the list is freed and its ptr set to ```nullptr```. Further ```addNewDSOImage``` calls will then just load the section data for the cache rather than storing the address.

The per cache mutex is used in case a new DSO is loaded whilst the current DSOs are being processed.

Does this make sense?

Comment (jckarter) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: I see. If that's the intended design, then `swift_once` here isn't giving any benefit over a global constructor, since the entry point will be called at load time in almost any real program. I still think we should avoiding doing any work until the caches are used; it delays start time and wastes effort in a program that never uses the lookup mechanisms.

Comment (spevans) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: @gparker If it is statically linked then ```addNewDSOImage()``` is a nop and the cache initialise functions just access the sections directly, see https://github.com/apple/swift/blob/master/stdlib/public/runtime/ImageInspectionStatic.cpp

**Comment by slavapestov** [comment]
@jckarter Can you take another look at this? Thanks!

**Comment by spevans** [comment]
@jckarter Ive had another look at this and used your suggestion for the constructor of just testing a 'lookup initialized' flag for each cache. This keeps ```addNewDSOImage()``` as lazy and minimal as possible.

I also fixed a bug with the ```dl_iterate_phdr()``` callback where it gets called for both the dynamic loader and the main executable with a filename of ```NULL``` so was loading the data in the executable twice.

File: stdlib/public/runtime/ImageInspection.h [hunk 4]
```diff
@@ -30,6 +30,18 @@ namespace swift {
...
 
+#if defined(__ELF__) || defined(__ANDROID__)
+
+struct SectionInfo {
+  uint64_t size;
+  const char *data;
+};
+
+// Called by injected constructors when a dynamic library is loaded.
+void addNewDSOImage(const void *addr);
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspection.h]: Since this is exported from the runtime, we should give it an `extern "C"` name like `swift_addNewDSOImage`, so that the runtime ABI isn't dependent on the C++ ABI.

File: stdlib/public/runtime/ImageInspection.h [hunk 5]
```diff
@@ -30,6 +30,18 @@ namespace swift {
     void *symbolAddress;
   };
 
+#if defined(__ELF__) || defined(__ANDROID__)
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspection.h]: Could we separate this #if block out into a separate `ImageInspectionELF.h` header?

File: stdlib/public/runtime/ImageInspectionStatic.cpp [hunk 6]
```diff
@@ -14,22 +14,24 @@
...
-struct SectionInfo {
-  uint64_t size;
-  const char *data;
-};
+using namespace swift;
+
+// Called from ImageInspectionInit
+void
+swift::addNewDSOImage(const void *addr) {
+}
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspectionStatic.cpp]: This shouldn't have to exist at all in a static executable. I'd rather we get a link error if we accidentally try to link in the static constructor stub than quietly link in a useless static constructor.

File: stdlib/public/runtime/ImageInspectionELF.cpp [hunk 7]
```diff
@@ -35,67 +37,124 @@ static const char ProtocolConformancesSymbol[] =
...
+                               size_t size, void *data) {
+  const InspectArgs *inspectArgs = reinterpret_cast<const InspectArgs *>(data);
+  const char *fname = info->dlpi_name;
+
+  // While dl_iterate_phdr() is in progress it holds a lock to prevent other
+  // images being loaded. The initialize flag is set here inside the callback so
+  // that addNewDSOImage() sees a consistent state. If it was set outside the
+  // dl_interate_phdr() call then it could result in images being missed or
+  // added twice.
+  inspectArgs->didInitializeLookup->store(true, std::memory_order_release);
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: If it's guaranteed that `dlopen` and `dl_iterate_phdr` hold the same lock, then the `didInitializeLookup` flag should be able to be nonatomic.

File: stdlib/public/runtime/ImageInspectionELF.cpp [hunk 8]
```diff
@@ -35,67 +37,124 @@ static const char ProtocolConformancesSymbol[] =
...
-    reinterpret_cast<const char*>(dlsym(handle, inspectArgs->symbolName));
 
-  if (!conformances) {
-    // if there are no conformances, don't hold this handle open.
+// Extract the section information for a named section in an image. imageName
+// can be nullptr to specify the main executable.
+static SectionInfo getSectionInfo(const char *imageName,
+                                  const char *sectionName) {
+  SectionInfo sectionInfo = { 0, nullptr };
+  void *handle = dlopen(imageName, RTLD_LAZY);
```
Comment (jckarter) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: Should this also pass `RTLD_NOLOAD`? The previous version did.

**Review by jckarter — COMMENTED** [review]
Looks good! Just a few minor requests:

File: stdlib/public/runtime/ImageInspection.h (hunk 4 above)
Comment (gparker42) [inline stdlib/public/runtime/ImageInspection.h]: After #6820 it's spelled `SWIFT_RUNTIME_EXPORT`. That handles all of the mangling and visibility fuss required to provide a symbol that is called by outside code.

**Comment by spevans** [comment]
@jckarter I have fixed all the minor requests and rebased.

I removed the ```RTLD_NOLOAD``` as I dont think it is necessary. ```dlopen(3)``` says it is only used to test if a shared object is already resident which should always be the case by the time the ```dl_iterate_phdr()``` calls the callback.

**Comment by jckarter** [comment]
@spevans What's the behavior if a different file is on disk at the same path from what was loaded if RTLD_NOLOAD isn't passed? Seems like a nice consistency check.

**Comment by spevans** [comment]
@jckarter If the file is already loaded then then the filesystem isnt examined. Since the filename comes from either ```dl_iterate_phdr()``` or ```dladdr()``` then it should already be loaded and in the link map.

However I see your point about it acting as an extra check, the returned handle should never be ```NULL``` in which case I think exiting with ```fatalError()``` may be the correct action, something like:

```
void *handle = dlopen(imageName, RTLD_LAZY | RTLD_NOLOAD);
if (!handle) {
  fatalError(/* flags = */ 0, "dlopen() failed on `%s': %s", imageName,
             dlerror());
} else {
  void *symbol = dlsym(handle, sectionName);
```

does this look ok?

**Comment by jckarter** [comment]
Sounds good. You could leave out the `else` since the nesting isn't necessary after the `if` errors out.

**Comment by spevans** [comment]
@jckarter Ive added  the ```RTLD_NOLOAD``` and extra error checking

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Review by jckarter — APPROVED** [review]

**Comment by jckarter** [comment]
Looks great!

**Comment by spevans** [comment]
Are the CI failures just temporary issues? The log seemed to show it being an issue in lldb?

**Comment by jckarter** [comment]
@spevans There's a known bug in the Jenkins Github plugin that it doesn't like rebases. You just need to trigger CI twice for it to get the right SHA, AIUI. (@shahmishal might have more background)

**Comment by jckarter** [comment]
I see the lldb failure now.  @egranata Are we disturbing the format of any metadata objects LLDB is looking at?

**Comment by jckarter** [comment]
Seems to be correlated, since the builds right before the ones I started for this PR passed…

**Comment by spevans** [comment]
OK, I'll take a look and see if I can replicate the test failure

**Comment by spevans** [comment]
@jckarter Looks like the same test failed in the last CI of 'Project OSS - Swift Package - Ubuntu 16.10 (master)' https://ci.swift.org/job/oss-swift-package-linux-ubuntu-16_10/328/

**Comment by jckarter** [comment]
OK, might be unrelated then. Let me try again.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by spevans** [comment]
@jckarter I think there is something broken in master for the linux tests as  https://ci.swift.org/job/oss-swift-package-linux-ubuntu-16_10 has been failing for the last day or so

**Comment by jckarter** [comment]
I agree. We can't merge additional PRs until master is clean, though. We'll have to wait for the lldb test to be fixed.

**Comment by jckarter** [comment]
Master seems to be clean now, let's try again.

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
Thanks again @spevans!
